torchelastic/distributed/launch.py

```diff
@@ -0,0 +1,369 @@
+3. Number of nodes is allowed to change (elasticity).
+
+**Usage:**
+
+1. Fault tolerant (no elasticity):
+
+::
+
+    >>> python -m torch.distributed.launch
+            --nnodes=4
+            --nproc_per_node=8
+            --rdzv_id=JOB_ID
+            --rdzv_backend=etcd
+            --rdzv_endpoint=ETCD_HOST:ETCD_PORT
+            --nproc_per_node=8
+            YOUR_TRAINING_SCRIPT.py (--arg1 ... train script args...)
+
+2. Elastic (min=1, max=4):
+
+::
+
+    >>> python -m torch.distributed.launch
+            --nnodes=1:4
+            --rdzv_id=JOB_ID
+            --rdzv_backend=etcd
+            --rdzv_endpoint=ETCD_HOST:ETCD_PORT
+            --nproc_per_node=8
+            YOUR_TRAINING_SCRIPT.py (--arg1 ... train script args...)
+
+**Definitions:**
+
+1. `Node` - Physical instance or container.
+Maps to the unit that the job manager works with.
+
+2. `Worker` - A worker in the context of distributed training.
+
+3. `Worker Group` - Workers with the same function (e.g. trainers)
+
+4. `Local Worker Group` - Subset of the workers in the
+worker group running on the same Node
+
+4. `RANK` - rank of the worker within a worker group.
+
+5. `WORLD_SIZE` - total number of workers in a worker group.
+
+6. `LOCAL_RANK` - rank of the worker within a local worker group
+
+7. `LOCAL_WORLD_SIZE` - size of the local worker group
+
+8. `rdzv_id` - user defined id that uniquely identifies the worker group
+for a job. This id is used by each node to join as a member of a particular
+worker group.
+
+9. `rdzv_backend` - the backend store of rendezvous (e.g. etcd). This is
+typically a strongly consistent key-value store.
+
+10. `rdzv_endpoint` - rdzv backend server endpoint in `host:port` format.
+
+A `Node` runs `LOCAL_WORLD_SIZE` workers which comprise a `LocalWorkerGroup`.
+The union of all `LocalWorkerGroups` in the nodes in the job comprise the
+`WorkerGroup`.
+
+**Deployment:**
+
+This launcher starts on agent process which creates and monitors
+a local worker group. For multi-node distributed training, simply
+start the launcher with the same arguments on all the nodes participating
+in training.
+
+When using a job/cluster manager the entry point command to the multi-node
+job is invoking this launcher.
+
+**Failure Modes:**
+
+1. Worker failure - For a training job with `n` workers, if `k < n` workers fail
+all workers are stopped and restarted up to `max_restarts`.
+
+2. Agent failure - An agent failure results in local worker group failure,
+it is up to the job manager to fail the entire job (gang semantics) or attempt
+to replace the node. Both behaviors are supported by the agent.
+
+3. Node failure - Same as agent failure.
+
+**Membership Changes:**
+
+1. Node departure (scale-down) - agent is notified of the departure,
+all existing workers are stopped, a new `Worker Group` is formed and all
+workers are started with a new `RANK` and `WORLD_SIZE`.
+
+2. Node arrival (scale-up) - the new node is admitted to the job,
```

Comment:
If the number of existing works have already surpassed the minimum `nnodes` but smaller than maximum `nnodes`, will node arrival cause existing workers to stop? Which components decides node admission?